Fix 304s and more.

michaellenaghan opened the pull request:
A quick summary:

 * Place double-quotes around ETag values.
 * Convert `str` bodies to `bytes`.
 * Always set the `content-length` header.
 * Return a 304 *after* `content-type` and `content-length` (and, maybe, `content-encoding`) headers have been set.
 * Don't set a "default" `content-type` header if the body is an empty byte string.
 * Don't set a "default" `content-length` header if the body is an empty byte string.

===

I created two temporary sites to illustrate some of the issues, and the fixes:

 * https://timetagger.nfshost.com/
 * https://timetaggerpatched.nfshost.com/

The first is an unpatched copy of TimeTagger. 

The second is a patched copy. 

Both are behind an Apache proxy.

To see the problem, go to https://timetagger.nfshost.com/timetagger/. Refresh a few times. You should (eventually?) get a 404. Refresh a few times. You should (eventually?) get a page again.

(In fact, it's often enough to just go to https://timetagger.nfshost.com/timetagger/ and hit the Home link in the top left corner.)

Apache seems to be unhappy. The most likely explanation is that the `content-type` and `content-length` change if a conditional request is made:

```bash
$ curl -v -H "If-None-Match: 512dc2f89fea2c3794631fddb080e9db842542b0204cd3e2b13d1685d9f2a349" "http://localhost:5555/timetagger/" -o /dev/null
...
< HTTP/1.1 304 Not Modified
< date: Tue, 05 Mar 2024 23:29:19 GMT
< server: uvicorn
< etag: 512dc2f89fea2c3794631fddb080e9db842542b0204cd3e2b13d1685d9f2a349
< cache-control: public, must-revalidate, max-age=0
< content-type: application/octet-stream
< content-length: 0
```

The code in `utils.py` doesn't set `content-type` and `content-length` on a conditional request — but the code in `_app.py` does. And they're different from the resources' original values, so Apache is right to be suspicious.

My original fix just moved the conditional check down, after the `content-type` and `content-length` were set in `utils.py`. I discovered that in that case the `content-length` was *still* different. That's because the content length was counting characters rather than bytes.

And once I fixed the issues in `utils.py` I discovered that the changes in `_app.py` were required. You can try that out yourself: only apply the `utils.py` changes to `asgineer` in `timetagger`, and then go back and forth between the Home and App links.

I tried to make minimal changes rather than elegant changes. I won't be hurt if you decide to go in a different direction. :-)

P.S. I'll take the test sites down in a few days.

michaellenaghan commented:
I forgot to mention one important thing: always setting `content-length` means that `transfer-encoding: chunked` is never used.

If that's a problem — and it may be for a single-threaded async network library? it probably depends on asset sizes — I can go back to setting it conditionally.

michaellenaghan commented on asgineer/utils.py:
```diff
@@ -138,12 +137,17 @@ async def asset_handler(request, path=None):
-            body = assets[path]
+            body = to_bytes(assets[path])
+
+        headers["content-length"] = str(len(body))
```
It was required before I made the change to `utils.py`; I'll see if that's no longer the case.

michaellenaghan replied:
OK, I think that will work. I'll update the PR.

Btw, here are the relevant parts of the relevant RFCs.

> ### [3.3.2.  Content-Length](https://www.rfc-editor.org/rfc/rfc7230#section-3.3.2)
>
> ...
>
> A server MAY send a Content-Length header field in a 304 (Not
   Modified) response to a conditional GET request (Section 4.1 of
   [RFC7232]); a server MUST NOT send Content-Length in such a response
   unless its field-value equals the decimal number of octets that would
   have been sent in the payload body of a 200 (OK) response to the same
   request.
>
> ...

> ## [4.1.  304 Not Modified](https://www.rfc-editor.org/rfc/rfc7232#section-4.1)
>
>   The 304 (Not Modified) status code indicates that a conditional GET
   or HEAD request has been received and would have resulted in a 200
   (OK) response if it were not for the fact that the condition
   evaluated to false.  In other words, there is no need for the server
   to transfer a representation of the target resource because the
   request indicates that the client, which made the request
   conditional, already has a valid representation; the server is
   therefore redirecting the client to make use of that stored
   representation as if it were the payload of a 200 (OK) response.
>
>   The server generating a 304 response MUST generate any of the
   following header fields that would have been sent in a 200 (OK)
   response to the same request: Cache-Control, Content-Location, Date,
   ETag, Expires, and Vary.
>
>   Since the goal of a 304 response is to minimize information transfer
   when the recipient already has one or more cached representations, a
   sender SHOULD NOT generate representation metadata other than the
   above listed fields unless said metadata exists for the purpose of
   guiding cache updates (e.g., Last-Modified might be useful if the
   response does not have an ETag field).
>
>   Requirements on a cache that receives a 304 response are defined in
   Section 4.3.4 of [RFC7234].  If the conditional request originated
   with an outbound client, such as a user agent with its own cache
   sending a conditional GET to a shared proxy, then the proxy SHOULD
   forward the 304 response to that client.
>
>   A 304 response cannot contain a message-body; it is always terminated
   by the first empty line after the header fields.

I'll revise accordingly.

michaellenaghan replied:
OK, this was a bigger change. My intention was to make the flow more top-to-bottom, and the decisions — i.e., removing certain headers from 304 responses — more obvious. If the changes make you sad, let me know. :-)

## Localhost

```bash
# 307 Response

$ curl -v "http://localhost:5555/" -o /dev/null
...
< HTTP/1.1 307 Temporary Redirect
< date: Fri, 08 Mar 2024 18:55:59 GMT
< server: uvicorn
< Location: /timetagger/
< Transfer-Encoding: chunked

# 200 Response

$ curl -v "http://localhost:5555/timetagger/" -o /dev/null
...
< HTTP/1.1 200 OK
< date: Fri, 08 Mar 2024 18:56:10 GMT
< server: uvicorn
< etag: "512dc2f89fea2c3794631fddb080e9db842542b0204cd3e2b13d1685d9f2a349"
< cache-control: public, must-revalidate, max-age=0
< content-type: text/html
< content-length: 9196
...

# 200 Response (HEAD Request)

$ curl -v -I "http://localhost:5555/timetagger/" -o /dev/null
 ...
< HTTP/1.1 200 OK
< date: Fri, 08 Mar 2024 18:56:26 GMT
< server: uvicorn
< etag: "512dc2f89fea2c3794631fddb080e9db842542b0204cd3e2b13d1685d9f2a349"
< cache-control: public, must-revalidate, max-age=0
< content-type: text/html
< content-length: 9196
...

# 304 Response

$ curl -v -H "If-None-Match: \"512dc2f89fea2c3794631fddb080e9db842542b0204cd3e2b13d1685d9f2a349\"" "http://localhost:5555/timetagger/" -o /dev/null
...
< HTTP/1.1 304 Not Modified
< date: Fri, 08 Mar 2024 18:56:37 GMT
< server: uvicorn
< etag: "512dc2f89fea2c3794631fddb080e9db842542b0204cd3e2b13d1685d9f2a349"
< cache-control: public, must-revalidate, max-age=0
...
```

## Server (behind Apache)

```bash
# 307 Response

$ curl -v "https://timetaggerpatched.nfshost.com/" -o /dev/null
 ...
< HTTP/2 307 
< date: Fri, 08 Mar 2024 19:00:30 GMT
< server: Apache
< location: /timetagger/
< vary: Accept-Encoding
< via: e1s
< content-length: 0
 ...

# 200 Response

$ curl -v "https://timetaggerpatched.nfshost.com/timetagger/" -o /dev/null
 ...
< HTTP/2 200 
< content-length: 9196
< date: Fri, 08 Mar 2024 19:00:39 GMT
< server: Apache
< cache-control: public, must-revalidate, max-age=0
< etag: "512dc2f89fea2c3794631fddb080e9db842542b0204cd3e2b13d1685d9f2a349"
< vary: Accept-Encoding
< via: e1s
< content-type: text/html
 ...

# 200 Response (HEAD Request)

$ curl -v -I "https://timetaggerpatched.nfshost.com/timetagger/" -o /dev/null
 ...
< HTTP/2 200 
< content-length: 9196
< date: Fri, 08 Mar 2024 19:00:50 GMT
< server: Apache
< cache-control: public, must-revalidate, max-age=0
< etag: "512dc2f89fea2c3794631fddb080e9db842542b0204cd3e2b13d1685d9f2a349"
< vary: Accept-Encoding
< via: e1s
< content-type: text/html
 ...

# 304 Response

$ curl -v -H "If-None-Match: \"512dc2f89fea2c3794631fddb080e9db842542b0204cd3e2b13d1685d9f2a349\"" "https://timetaggerpatched.nfshost.com/timetagger/" -o /dev/null
 ...
< HTTP/2 304 
< date: Fri, 08 Mar 2024 19:01:01 GMT
< server: Apache
< cache-control: public, must-revalidate, max-age=0
< etag: "512dc2f89fea2c3794631fddb080e9db842542b0204cd3e2b13d1685d9f2a349"
< vary: Accept-Encoding
< via: e1s
...
```

almarklein commented on asgineer/utils.py:
```diff
@@ -79,75 +79,82 @@ async def some_handler(request):
-        # Set content type
-        if path in ctypes:
-            headers["content-type"] = ctypes[path]
+        status = 200
+        headers = {}
+        headers["cache-control"] = f"public, must-revalidate, max-age={max_age:d}"
+        headers["content-length"] = str(len(unzipped[path]))
+        headers["content-type"] = ctypes[path]
+        headers["etag"] = f"\"{etags[path]}\""
+        body = unzipped[path]
```
```suggestion
        status = 200
        body = unzipped[path]
        headers = {}
        headers["cache-control"] = f"public, must-revalidate, max-age={max_age:d}"
        headers["content-length"] = str(len(body))
        headers["content-type"] = ctypes[path]
        headers["etag"] = f"\"{etags[path]}\""
```

almarklein replied:
mmm, the diff looks like a lot has changed, but I only moved the `body = ...` up, and then used `body` where content-length is set. Basically the same change as the other suggestion.

almarklein commented on asgineer/utils.py:
```diff
@@ -79,75 +79,82 @@ async def some_handler(request):
         # Get body, zip if we should and can
         if path in zipped and "gzip" in request.headers.get("accept-encoding", ""):
             headers["content-encoding"] = "gzip"
+            headers["content-length"] = str(len(zipped[path]))
             body = zipped[path]
```
```suggestion
            body = zipped[path]
            headers["content-length"] = str(len(body))
```

michaellenaghan replied:
Done, but I've spotted another issue; investigating...

almarklein reviewed (commented):
I like it!

I made two small suggestions to avoid a duplicate dict lookup and make the code a bit easier to read.

michaellenaghan commented:
Actually, I think I've spotted *two* more issues. 

Here's the first. 

Take a look at this:

```bash
$ curl -v "http://localhost:5555/"
...
> GET / HTTP/1.1
> Host: localhost:5555
> User-Agent: curl/8.4.0
> Accept: */*
> 
< HTTP/1.1 307 Temporary Redirect
< date: Mon, 11 Mar 2024 16:25:50 GMT
< server: uvicorn
< Location: /timetagger/
< Transfer-Encoding: chunked
< 
* Connection #0 to host localhost left intact
```

The response is marked as `Transfer-Encoding: chunked`, but no chunks follow. I think that's wrong. I think there should be at least:

```
0\r\n
\r\n
```

Otherwise, how will the client be able to determine where one message ends and the next message begins?

The relevant part of the spec is [here](https://www.rfc-editor.org/rfc/rfc7230#section-4.1). 

There are examples [here](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Transfer-Encoding).

This particular 307 comes from `timetagger` rather than `asgineer`, [here](https://github.com/almarklein/timetagger/blob/1eb0dc25e0a1d7a3a9267be978e50c1b2667e5e0/timetagger/__main__.py#L86).

michaellenaghan commented:
It's been difficult to figure out whether or not the 307 response *really* has no chunks, or if `curl` just doesn't show chunks — even when it's logging verbosely.

I'm actually more interested in the *second* issue: if I click the Home link repeatedly in the patched TimeTagger test site I can induce errors. I don't know yet whether the errors are real (and due to `asgineer`) or somehow just a byproduct of my hosting environment.

I don't want to dismiss the errors too quickly, though; we already found some problems, and there could plausibly be others.

P.S. I know that the *first* issue, the 307 issue, is indirectly related to one of my `_app.py` changes. I'll explain later; I'm still thinking about options.

michaellenaghan commented:
OK, I think the *second* issue, the sporadic errors, are "somehow just a byproduct of my hosting environment."

And after thinking about the *first* issue a bit more, I've changed the `_app.py` code.

===

Originally, `_app.py` had this:

```python
    ...
    headers.setdefault("content-length", str(len(body)))
    ...
```

I changed it to this:

```python
    ...
    if body != b"":
        headers.setdefault("content-length", str(len(body)))
    ...
```

I made the change because, without it, the 304 had a `Content-Length` of 0.

Now, with it, the 307 no longer has a `Content-Length` header — and so `uvicorn` is [adding](https://github.com/encode/uvicorn/blob/4a503d84fa8703d7534d810bb10b3a0b0e6e1a39/uvicorn/protocols/http/httptools_impl.py#L516) a `Transfer-Encoding: chunked` header.

A `Transfer-Encoding: chunked` header with an empty body doesn't feel quite right.

===

I considered four options:

 * Leave it; it's *odd*, but it isn't necessarily *wrong*.
 
 * Patch it, take 1; add a `content-length: 0` header to the redirect in `timetagger`.

 * Patch it, take 2; back out the original `_app.py` change and go back to sending the unnecessary-but-not-wrong 304 headers (`content-encoding`, `content-length`, `content-type`) in `utils.py`.

 * Fix it; rethink the `_app.py` changes.

This is my take on the "fix it" option.

It's loosely based on [this](https://github.com/encode/uvicorn/blob/4a503d84fa8703d7534d810bb10b3a0b0e6e1a39/uvicorn/protocols/http/httptools_impl.py#L513) line in `uvicorn` and on [this](https://www.rfc-editor.org/rfc/rfc7230#section-3.3) section of RFC 7230.

I think the general direction is good, but you may want some stylistic changes.

P.S. To answer an obvious potential objection: I'm passing two unused params, `headers` and `body`, to the new `_response_has_body()` method because I want to "hide" the basis for the decision-making. The code I just removed, for example, special-cased on empty (`b""`) bodies.

almarklein commented:
Thanks!

michaellenaghan commented:
@almarklein Could you bump the version at some point so that this fix makes its way out into the world?

almarklein commented:
Done!

I was just thinking, it would also be nice to have unit tests to cover these cases (in `tests/test_utils.py`). That way we can also make sure that the behavior is maintained if other changes are made in the future. Maybe for another day ...
